### DISCUSSION TITLE
How to redirect user outside of React component?

### DISCUSSION THREAD

**Discussion by hazratgs** [vercel/next.js#57495] [App Router]
I'm migrating a project from CRA to Next.js 13 and the code often uses the `history.push('/')` methods and the `react-router` library, something similar is in Pages Router:

```
import Router from 'next/router'

.  .  .

Router.push('/')
```

But there is no such approach in **App Router**, only hooks. **Is there a way to use the router outside of a React application?**

**Comment by Hillsie** [comment]
It's  in `next/navigation`;

https://nextjs.org/docs/app/api-reference/functions/use-router


```
import {  useRouter } from "next/navigation";
...
 const router = useRouter();
 ....
 router.push('/');
````

**Reply by hazratgs** [reply]
I read the documentation, but I don't need a hook, but the ability to execute .push() outside of React Component

for example, the project uses redux-saga and the logic needs to switch to another page

**Comment by landsman** [comment]
register it as callback to window object and call it from whatever you want :)

**Comment by aziyatali** [comment]
`
import { NextResponse } from 'next/server';
`
You can do something like this

`
const some_url = process.env.someenv; (or any custom url)
`
`return NextResponse.redirect(new URL(some_url));
`

Can do the same in client-side

**Comment by michal-th** [comment]
I'd like to implement redirection logic in my client fetch wrapper  (intercepting some specific 4xx status codes that provide redirect url as payload),  can't figure out any clean way of accessing the client side router - currently had to resolve it with a dirty workaround using window.location..

**Comment by barman47** [comment]
I found this page yesterday when I had a similar issue. `window.location.href` was not ideal for me because it messed up my redux state and I had code running before the state was hydrated. I was able to get it to work so I thought to share in case anyone still needed help.

```
'use server';

import { redirect } from 'next/navigation';

export async function navigate(url: string) {
    redirect(url);
}
```
My usage:

```
import { navigate } from '@/app/actions';
. . . 
navigate(`/signUpSuccess?email=${email}`);
```

This is also contained in the docs.

**Reply by MaxKoldun** [reply]
thx it helped me

**Reply by Slava-Nik** [reply]
It helped, thank you.

But generally, is using this server-side code on the client side okay in this case?

I wonder why Next.js stopped providing such a basic and simple functionality? 
Yes, I want move my business logic to sagas or somewhere else, and I don't want to keep it in components. 

Before this code worked well in Next 12. 

import Router from 'next/router'
.  .  .
Router.push('/')

But now we are getting strange errors like "you should use next/router on client side". (It's client side already, not hooks though.)

**Comment by RSkrypnik** [comment]
Omg, I cannot believe it's still a problem in next/react unlike nuxt/vue where you can redirect outside of components

**Comment by q010p** [comment]
Just a heads-up for anyone else running into this:

When you use router.push('/api/your-endpoint'), it first tries to fetch the route via client-side navigation (using fetch). But because it's an API route and not a Next.js page, the browser falls back to a full page reload (document). As a result, the API endpoint gets hit twice, which can be super confusing and frustrating to debug.

**Comment by akegvd** [comment]
try this bro
I use this implement with Nextjs 15/16 app router with output=export (client side only)
it work well

```typescript
// src/modules/globalRouter/utils/globalRouter.ts
import type { AppRouterInstance } from 'next/dist/shared/lib/app-router-context.shared-runtime';

/**
 * Global router reference for use outside React components (e.g. axios interceptors).
 * Assigned by GlobalRouterSetup in the root layout so navigation works in non-component code.
 * see the initialize in src/modules/globalRouter/components/GlobalRouterSetup.tsx
 */
type TGlobalRouter = Pick<AppRouterInstance, 'push' | 'replace'>;

const globalRouter: TGlobalRouter = {
  push: () => {
    throw new Error('Global router not initialized');
  },
  replace: () => {
    throw new Error('Global router not initialized');
  },
};

export default globalRouter;
```

```typescript
// src/modules/globalRouter/components/GlobalRouterSetup.tsx
'use client';

import { useRouter } from 'next/navigation';
import { useEffect } from 'react';

import globalRouter from '../utils/globalRouter';

export default function GlobalRouterSetup() {
  const router = useRouter();

  useEffect(() => {
    globalRouter.push = router.push;
    globalRouter.replace = router.replace;
  }, [router]);

  return null;
}
```

```typescript
// src/app/layout.tsx (root layout)
'use client';

import React from 'react';

import GlobalRouterSetup from '@/app/GlobalRouterSetup';

export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <html lang="en">
      <body>
        <GlobalRouterSetup />
        {children}
      </body>
    </html>
  );
}

```

for usage
```typescript
// .../interceptor.ts axios I do
import globalRouter from '@/modules/globalRouter/utils/globalRouter';

const onResponseError = async (error: AxiosError<IApiErrorResponse | undefined>): Promise<never> => {
  if (error.code === 'ERR_CANCELED') {
    return Promise.reject(new Error(REQUEST_CANCELED_MESSAGE));
  }

  // NOTE: edge case, no response from server
  if (!error.response) {
    globalRouter.replace(ROUTES.error);
    return Promise.reject(new Error(NO_RESPONSE_MESSAGE));
  }

 ....
};
```
